src/update-index.js

```diff
@@ -43,37 +53,81 @@ const error = (path, e) => ({
 /**
  * Find an existing entry, given its path or sourceHash and branch.
  *
- * @param {object} index Algolia index
+ * @param {SearchIndex} index Algolia index
  * @param {object} attributes Attributes to search for
- * @param {string} branch Branch
  */
 async function search(index, attributes) {
-  const filters = Object.getOwnPropertyNames(attributes).map(
-    (name) => `${name}:${attributes[name]}`,
-  );
+  const filters = Object.getOwnPropertyNames(attributes)
+    .filter(
+      (name) => attributes[name],
+    )
+    .map(
+      (name) => `${name}:${attributes[name]}`,
+    );
   const searchresult = await index.search({
     attributesToRetrieve: ['path', 'name', 'objectID', 'sourceHash'],
     filters: filters.join(' AND '),
   });
   return searchresult.nbHits !== 0 ? searchresult.hits[0] : null;
 }
 
+/**
+ * Prepare items to be re-indexed.
+ *
+ * @param {object} cfg index configuration
+ * @param {SearchIndex} index algolia index
+ * @param {object} coll collection of items
+ * @param {string} branch GitHub branch
+ *
+ * @return {Array} of { path, hit } items
+ */
+async function prepareItems(cfg, index, coll, branch) {
+  let searchresults = (await Promise.all(coll.items.map(
+    // try to lookup path for items that only have a source hash
+    async (item) => {
+      const { path, sourceHash } = item;
+      const hit = await search(index, { path, sourceHash, branch });
+      return { path: path || (hit ? hit.path : null), hit };
+    },
+  ))).filter(
+    // drop tems that still haven't a valid path
+    ({ path }) => !!path,
+  );
+  const { mountpoint } = coll;
+  if (mountpoint) {
+    // if mountpoint is given, translate paths and remove leading slash
+    const re = new RegExp(`^${mountpoint.root}/`);
+    const repl = mountpoint.path.replace(/^\/+(.*)/, '$1');
+
+    searchresults = searchresults.map(
+      (item) => ({ path: item.path.replace(re, repl), hit: item.hit }),
+    );
+  }
+  return searchresults.filter(
+    // keep only items that are included in the index definition
+    ({ path }) => includes(index, path),
+  ).map(({ path, hit }) => {
+    // replace requested extension with the one in source
+    const noext = path.replace(/([^.]+)\.[^./]+/, '$1');
+    return { path: `${noext}.${cfg.source}`, hit };
+  });
+}
+
 /**
  * Update all records in a index with the paths given.
  *
  * @param {object} params parameters
- * @param {object} name index name
+ * @param {object} cfg index configuration
  * @param {SearchIndex} index algolia index
+ * @param {object} coll collection of items to index
  * @returns HTTP multi response
  */
-module.exports = async (params, name, index, paths) => {
+module.exports = async (params, cfg, index, coll) => {
```

Comment:
our convention is to never export anonymous functions.
better to declare the function and then export it:

```
async function updateIndex() { ... }
...
module.exports = updateIndex;
```

Reply:
I didn't know that: to me, naming a file `update-index.js` and exporting *one* function from it makes its purpose pretty obvious, but I'll change it.

Reply:
the main reason is to get nicer stack traces.